packages/aws-cdk/lib/api/hotswap/s3-bucket-deployments.ts

```diff
@@ -0,0 +1,107 @@
+import { ISDK } from '../aws-auth';
+import { ChangeHotswapImpact, ChangeHotswapResult, HotswapOperation, HotswappableChangeCandidate, establishResourcePhysicalName } from './common';
+import { EvaluateCloudFormationTemplate } from './evaluate-cloudformation-template';
+
+/**
+ * This means that the value is required to exist by CloudFormation's API (or our S3 Bucket Deployment Lambda)
+ * but the actual value specified is irrelevant
+ */
+export const REQUIRED_BY_CFN = 'required-to-be-present-by-cfn';
+
+export async function isHotswappableS3BucketDeploymentChange(
+  logicalId: string, change: HotswappableChangeCandidate, evaluateCfnTemplate: EvaluateCloudFormationTemplate,
+): Promise<ChangeHotswapResult> {
+  // In old-style synthesis, the policy used by the lambda to copy assets Ref's the assets directly,
+  // meaning that the changes made to the Policy are artifacts that can be safely ignored
+  if (change.newValue.Type === 'AWS::IAM::Policy') {
+    const roles = change.newValue.Properties?.Roles;
+    for (const role of roles) {
+      const roleLogicalName = await evaluateCfnTemplate.findLogicallNameFor(await evaluateCfnTemplate.evaluateCfnExpression(role));
+      if (!roleLogicalName) {
+        return ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT;
+      }
+
+      const roleRefs = evaluateCfnTemplate.findReferencesTo(roleLogicalName);
+      for (const roleRef of roleRefs) {
+        if (roleRef.Type === 'AWS::Lambda::Function') {
+          const lambdaRefs = evaluateCfnTemplate.findReferencesTo(roleRef.LogicalId);
+          for (const lambdaRef of lambdaRefs) {
+            // If S3Deployment -> Lambda -> Role and IAM::Policy -> Role, then this IAM::Policy change is an
+            // artifact of old-style synthesis
+            if (lambdaRef.Type === 'Custom::CDKBucketDeployment') {
+              return new EmptyHotswapOperation();
+            }
+          }
```

Comment:
I think the logic here is the wrong-way round. If we _only_ find Policies that _only_ reference Roles that _only_ reference a Lambda that _only_ references a `Custom::CDKBucketDeployment`, _then_ we can return `EmptyHotswapOperation`.

If that gigantic 'and' expression from above is not true (for any reason), then we need to return `ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT`.

So, the loop should check `if (roleRef.Type !== 'AWS::Lambda::Function') return ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT;`, and the same for the other conditions. And then, if we don't early return, then you can return `new EmptyHotswapOperation()`.

This will probably require extracting this to a helper function, but that's probably a good idea either way.